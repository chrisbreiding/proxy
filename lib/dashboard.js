const garage = require('./garage')
const notion = require('./notion')
const weather = require('./weather')

const wrap = (who, fn) => {
  return fn()
  .then((data) => {
    return { data }
  })
  .catch((error) => {
    // eslint-disable-next-line no-console
    console.log('Getting', who, 'data errored:', error.stack)

    return {
      name: error.name,
      message: error.message,
      stack: error.stack,
    }
  })
}

const get = (req, res) => {
  Promise.all([
    wrap('garage', () => garage.getData()),
    wrap('notion', () => notion.getData(req.query)),
    wrap('weather', () => weather.getData(req.query)),
  ])
  .then(([garage, notion, weather]) => {
    res.json({ garage, notion, weather })
  })
}

module.exports = {
  get,
}
