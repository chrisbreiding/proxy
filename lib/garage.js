const fs = require('fs-extra')
const rp = require('request-promise')

// In production, this is mounted with dokku's persistent storage
// https://github.com/dokku/dokku/blob/master/docs/advanced-usage/persistent-storage.md
const dataPath = process.env.NODE_ENV === 'development' ?
  './garage-data.json' :
  '/storage/garage-data.json'

const errorResult = {
  left: 'unknown',
  right: 'unknown',
}

const getData = () => fs.readJSON(dataPath)

const get = (req, res) => {
  getData()
  .then((data) => {
    if (data) {
      res.json(data)
    } else {
      res.json(errorResult)
    }
  })
  .catch(() => {
    res.json(errorResult)
  })
}

const notify = (message) => {
  const query = `?value1=${encodeURIComponent(message)}`
  const url = `https://maker.ifttt.com/trigger/notify/with/key/${process.env.IFTTT_WEBHOOK_KEY}${query}`

  return rp(url)
  .catch((error) => {
    // eslint-disable-next-line no-console
    console.log('IFTTT webhook request errored:', error.stack)
  })
}

const set = (req, res) => {
  const { door, state } = req.params

  getData()
  .then((data) => {
    const previousState = data[door]

    if (previousState === 'open' && state === 'open') {
      // if the door state was 'open' twice in a row, the door failed to close.
      // send a request to IFTTT, which will trigger a push notification.
      notify(`The ${door} garage door failed to close!`)
    } else if (state === 'open' && data.notifyOnOpen) {
      notify(`The ${door} garage door opened`)
    }

    const update = {
      [`${door}Previous`]: previousState,
      [door]: state,
    }

    return fs.outputJSON(dataPath, Object.assign(data, update))
  })
  .catch((error) => {
    // eslint-disable-next-line no-console
    console.log(`Setting garage state (${door} => ${state}) errored:`, error.stack)
  })
  .finally(() => {
    res.json({})
  })
}

const setNotifyOnOpen = (req, res) => {
  const notifyOnOpen = req.params.notifyOnOpen === 'true'

  getData()
  .then((data) => {
    return fs.outputJSON(dataPath, Object.assign(data, { notifyOnOpen }))
  })
  .catch((error) => {
    // eslint-disable-next-line no-console
    console.log(`Setting notifyOnOpen to ${notifyOnOpen} errored:`, error.stack)
  })
  .finally(() => {
    res.json({})
  })
}

const view = (req, res) => {
  res.set('Content-Type', 'text/html')

  getData()
  .then((data) => {
    res.render('garages', { states: data, layout: false })
  })
  .catch((error) => {
    res.render('error', { message: error.message, stack: error.stack, layout: false })
  })
}

module.exports = {
  get,
  set,
  setNotifyOnOpen,
  view,
}
