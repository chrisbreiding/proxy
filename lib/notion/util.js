const rp = require('request-promise')

const makeRequest = ({ notionToken, body, method = 'get', path }) => {
  return rp({
    method,
    body,
    uri: `https://api.notion.com/v1/${path}`,
    headers: {
      'Authorization': `Bearer ${notionToken}`,
      'Notion-Version': '2021-05-13',
    },
    json: true,
  })
}

const getBlockChildren = ({ notionToken, pageId }) => {
  return makeRequest({ notionToken, path: `blocks/${pageId}/children` })
}

const appendBlockChildren = ({ notionToken, pageId, blocks }) => {
  return makeRequest({
    notionToken,
    method: 'patch',
    path: `blocks/${pageId}/children`,
    body: {
      children: blocks,
    },
  })
}

const updateBlock = ({ notionToken, blockId, block }) => {
  return makeRequest({
    notionToken,
    method: 'patch',
    path: `blocks/${blockId}`,
    body: block,
  })
}

const getPlainText = (withText) => {
  if (!withText || !withText.text) return

  return withText.text
  .map(({ plain_text }) => plain_text)
  .join('')
  .trim()
}

const weatherIconMap = {
  'clear-day': '☀️',
  'clear-night': '☀️',
  'rain': '☔️',
  'snow': '❄️',
  'sleet': '🌨',
  'wind': '💨',
  'fog': '🌫',
  'cloudy': '☁️',
  'partly-cloudy-day': '⛅️',
  'partly-cloudy-night': '⛅️',
  'default': '🌑',
}

const getWeatherIcon = (iconName) => {
  return weatherIconMap[iconName] || weatherIconMap.default
}

const makeTextPart = (content, color) => {
  const textPart = { text: { content } }

  if (color) {
    textPart.annotations = { color }
  }

  return textPart
}

module.exports = {
  getBlockChildren,
  appendBlockChildren,
  updateBlock,
  getPlainText,
  getWeatherIcon,
  makeTextPart,
}
