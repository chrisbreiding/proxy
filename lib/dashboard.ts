import type express from 'express'

import { getGarageData } from './garage'
import { getNotionData } from './notion'
import { getWeatherData } from './weather'

async function wrap (who: string, fn: () => Promise<any>) {
  try {
    const data = await fn()

    return { data }
  } catch (error: any) {
    // eslint-disable-next-line no-console
    console.log('Getting', who, 'data errored:', error.stack)

    return {
      name: error.name,
      message: error.message,
      stack: error.stack,
    }
  }
}

export async function get (req: express.Request, res: express.Response) {
  const { location, notionToken, notionPageId } = req.query

  if (!location || typeof location !== 'string') {
    return res.json({ error: 'Must include location in query' })
  }
  if (!notionToken || typeof notionToken !== 'string') {
    return res.json({ error: 'Must include notionToken in query' })
  }
  if (!notionPageId || typeof notionPageId !== 'string') {
    return res.json({ error: 'Must include notionPageId in query' })
  }

  const [garage, notion, weather] = await Promise.all([
    wrap('garage', () => getGarageData()),
    wrap('notion', () => getNotionData({ notionToken, notionPageId })),
    wrap('weather', () => getWeatherData({ location })),
  ])

  res.json({ garage, notion, weather })
}
