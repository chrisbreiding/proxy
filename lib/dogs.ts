import type express from 'express'
import { request } from './util/network'

export async function get (req: express.Request, res: express.Response) {
  try {
    const response = await request({
      url: `https://prod-hal-api.homeatlastdogrescue.com/dogs/${req.params.id}`,
    })

    res.json(response)
  } catch (error: any) {
    res.status(500).json({
      error,
      data: error?.response?.data,
    })
  }
}

export async function search (req: express.Request, res: express.Response) {
  try {
    const response = await request({
      url: 'https://prod-hal-api.homeatlastdogrescue.com/dogs/search',
      params: req.query,
    })

    res.json(response)
  } catch (error: any) {
    res.status(500).json({
      error,
      data: error?.response?.data,
    })
  }
}
