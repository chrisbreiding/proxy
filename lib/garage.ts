import type express from 'express'

import { debug } from './util/debug'
import { request } from './util/network'
import { GarageState, PersistentData } from './util/persistent-data'

const defaultData = {
  left: 'unknown',
  right: 'unknown',
}

const persistentData = new PersistentData('garage-data')

export async function get (req: express.Request, res: express.Response) {
  try {
    const data = await persistentData.get()

    if (data) {
      res.json(Object.assign({}, defaultData, data))
    } else {
      res.json(defaultData)
    }
  } catch (error) {
    res.json(defaultData)
  }
}

async function notify (message: string) {
  const query = `?value1=${encodeURIComponent(message)}`
  const url = `https://maker.ifttt.com/trigger/notify/with/key/${process.env.IFTTT_WEBHOOK_KEY}${query}`

  try {
    await request({ url })
  } catch (error: any) {
    /* c8 ignore next */
    debug('IFTTT webhook request errored:', error?.stack || error)
  }
}

type Door = 'left' | 'right'

export async function set (req: express.Request, res: express.Response) {
  const { door, state } = req.params as { door: Door, state: GarageState }

  try {
    const data = (await persistentData.get()) || {}
    const previousState = data[door]

    if (previousState === 'open' && state === 'open') {
      // if the door state was 'open' twice in a row, the door failed to close.
      // send a request to IFTTT, which will trigger a push notification.
      notify(`The ${door} garage door failed to close!`)
    } else if (state === 'open' && data.notifyOnOpen) {
      notify(`The ${door} garage door opened`)
    }

    const update = {
      [door]: state,
    }

    await persistentData.set(Object.assign({}, data, update))
  } catch (error: any) {
    /* c8 ignore next */
    debug(`Setting garage state (${door} => ${state}) errored:`, error?.stack || error)
  } finally {
    res.json({})
  }
}

export async function setNotifyOnOpen (req: express.Request, res: express.Response) {
  const notifyOnOpen = req.params.notifyOnOpen === 'true'

  try {
    const data = (await persistentData.get()) || {}

    await persistentData.set(Object.assign(data, { notifyOnOpen }))
  } catch (error: any) {
    /* c8 ignore next */
    debug(`Setting notifyOnOpen to ${notifyOnOpen} errored:`, error?.stack || error)
  } finally {
    res.json({})
  }
}

export async function view (req: express.Request, res: express.Response) {
  res.set('Content-Type', 'text/html')

  try {
    const data = Object.assign({}, defaultData, (await persistentData.get()) || {})

    res.render('garages', { states: data, layout: false })
  } catch (error: any) {
    res.render('error', { message: error.message, stack: error.stack, layout: false })
  }
}

export const getGarageData = persistentData.get.bind(persistentData)
