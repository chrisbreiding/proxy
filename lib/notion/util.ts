import type { Method } from 'axios'
import dayjs from 'dayjs'
import type {
  ApiColor,
  BlockObjectResponse,
  BookmarkBlockObjectResponse,
  BulletedListItemBlockObjectResponse,
  CalloutBlockObjectResponse,
  ChildDatabaseBlockObjectResponse,
  ChildPageBlockObjectResponse,
  ColumnBlockObjectResponse,
  ColumnListBlockObjectResponse,
  DividerBlockObjectResponse,
  EmbedBlockObjectResponse,
  EquationBlockObjectResponse,
  FileBlockObjectResponse,
  Heading1BlockObjectResponse,
  Heading2BlockObjectResponse,
  Heading3BlockObjectResponse,
  ImageBlockObjectResponse,
  LinkPreviewBlockObjectResponse,
  LinkToPageBlockObjectResponse,
  ListBlockChildrenResponse,
  ListDatabasesResponse,
  NumberedListItemBlockObjectResponse,
  ParagraphBlockObjectResponse,
  PdfBlockObjectResponse,
  QuoteBlockObjectResponse,
  RichTextItemResponse,
  SyncedBlockBlockObjectResponse,
  TableBlockObjectResponse,
  TableOfContentsBlockObjectResponse,
  TableRowBlockObjectResponse,
  TemplateBlockObjectResponse,
  ToDoBlockObjectResponse,
  ToggleBlockObjectResponse,
  UnsupportedBlockObjectResponse,
  UpdateBlockBodyParameters,
  VideoBlockObjectResponse,
} from '@notionhq/client/build/src/api-endpoints'

import { clone, compact } from '../util/collections'
import { debug } from '../util/debug'
import { request } from '../util/network'

export const dateRegex = /(?:Sun|Mon|Tue|Wed|Thu|Fri|Sat), (\d{1,2}\/\d{1,2})/

export function getDateFromText (dateText: string) {
  const currentDate = dayjs()
  // originally assume the date's year matches the current year
  const date = dayjs(`${dateText}/${currentDate.year()}`, 'M/D/YYYY')
  // if the current month is after the date's month, we've crossed over years
  // and the date's year should be the next year. usually, it's because it's
  // currently December, but the date's month is January. this can be
  // generically applied, so it will work if it's currently June, but we're
  // trying to get the weather for April. we don't want past weather, only
  // future weather, so assume it's for the following year.
  if (currentDate.month() > date.month()) {
    return date.add(1, 'year')
  }

  return date
}

interface MakeRequestOptions {
  body?: object
  method?: Method
  notionToken: string
  path: string
}

function makeRequest<T> (options: MakeRequestOptions): Promise<T> {
  const { notionToken, body, path, method = 'get' } = options

  return request({
    method,
    body,
    url: `https://api.notion.com/v1/${path}`,
    headers: {
      'Authorization': `Bearer ${notionToken}`,
      'Notion-Version': '2022-06-28',
    },
  })
}

interface GetBlockChildrenOptions {
  notionToken: string
  pageId: string
}

export async function getBlockChildren ({ notionToken, pageId }: GetBlockChildrenOptions) {
  const { results } = await makeRequest<ListBlockChildrenResponse>({ notionToken, path: `blocks/${pageId}/children` }) as { results: BlockObjectResponse[] }

  return results.map(convertNotionBlockToOwnBlock)
}

interface GetBlockChildrenDeepOptions {
  filter?: (block: NotionBlock) => boolean
  includeId?: boolean
  notionToken: string
  pageId: string
}

export async function getBlockChildrenDeep ({ notionToken, pageId, filter, includeId }: GetBlockChildrenDeepOptions): Promise<NotionBlock[]> {
  const blocks = await getBlockChildren({ notionToken, pageId })
  const filteredBlocks = filter ? blocks.filter(filter) : blocks
  const blocksWithChildren = await Promise.all(filteredBlocks.map(async (block) => {
    if (block.has_children && block.type !== 'child_page') {
      block.children = await getBlockChildrenDeep({ notionToken, pageId: block.id, includeId })
    }

    return block
  }))

  return compact(blocksWithChildren)
}

type Content = BookmarkBlockObjectResponse['bookmark']
| BulletedListItemBlockObjectResponse['bulleted_list_item']
| CalloutBlockObjectResponse['callout']
| ChildDatabaseBlockObjectResponse['child_database']
| ChildPageBlockObjectResponse['child_page']
| ColumnBlockObjectResponse['column']
| ColumnListBlockObjectResponse['column_list']
| DividerBlockObjectResponse['divider']
| EmbedBlockObjectResponse['embed']
| EquationBlockObjectResponse['equation']
| FileBlockObjectResponse['file']
| Heading1BlockObjectResponse['heading_1']
| Heading2BlockObjectResponse['heading_2']
| Heading3BlockObjectResponse['heading_3']
| ImageBlockObjectResponse['image']
| LinkPreviewBlockObjectResponse['link_preview']
| LinkToPageBlockObjectResponse['link_to_page']
| NumberedListItemBlockObjectResponse['numbered_list_item']
| ParagraphBlockObjectResponse['paragraph']
| PdfBlockObjectResponse['pdf']
| QuoteBlockObjectResponse['quote']
| SyncedBlockBlockObjectResponse['synced_block']
| TableBlockObjectResponse['table']
| TableOfContentsBlockObjectResponse['table_of_contents']
| TableRowBlockObjectResponse['table_row']
| TemplateBlockObjectResponse['template']
| ToDoBlockObjectResponse['to_do']
| ToggleBlockObjectResponse['toggle']
| UnsupportedBlockObjectResponse['unsupported']
| VideoBlockObjectResponse['video']

export interface OwnBlock {
  children?: OwnBlock[] | NotionBlock[]
  type: BlockObjectResponse['type']
  content: Content
}

export interface NotionBlock extends OwnBlock {
  children?: NotionBlock[]
  has_children: boolean
  id: string
}

export type Block = OwnBlock | NotionBlock

function getRichText (block: OwnBlock): RichTextItemResponse[] | undefined {
  return 'rich_text' in block.content ? block.content.rich_text : undefined
}

interface MakeAppendRequestOptions {
  blocks: OutgoingBlock[]
  notionToken: string
  pageId: string
}

export function makeAppendRequest ({ notionToken, pageId, blocks }: MakeAppendRequestOptions) {
  return makeRequest<ListBlockChildrenResponse>({
    notionToken,
    method: 'patch',
    path: `blocks/${pageId}/children`,
    body: {
      children: blocks,
    },
  })
  // don't understand why this fails converage
  /* c8 ignore next */
}

interface AppendBlockChildrenOptions {
  blocks: OwnBlock[]
  notionToken: string
  pageId: string
}

export async function appendBlockChildrenDeep ({ notionToken, pageId, blocks }: AppendBlockChildrenOptions) {
  function moveChildren (blocks: NotionBlock[] | OwnBlock[]) {
    return blocks.map((block) => {
      const convertedBlock = convertBlockToOutgoingBlock(block)

      if (!block.children) return convertedBlock

      // @ts-ignore
      convertedBlock[block.type].children = moveChildren(block.children)

      return convertedBlock
    })
  }

  return makeAppendRequest({ notionToken, pageId, blocks: moveChildren(blocks) })
}

export async function appendBlockChildren ({ notionToken, pageId, blocks }: AppendBlockChildrenOptions) {
  let toAppend: OutgoingBlock[] = []

  for (const block of blocks) {
    const convertedBlock = convertBlockToOutgoingBlock(block)

    if (block.children) {
      const children = block.children

      toAppend.push(convertedBlock)

      const { results } = await makeAppendRequest({ notionToken, pageId, blocks: toAppend })
      toAppend = []

      await appendBlockChildren({
        notionToken,
        pageId: results[results.length - 1].id,
        blocks: children,
      })
    } else {
      toAppend.push(convertedBlock)
    }
  }

  if (toAppend.length) {
    return makeAppendRequest({
      notionToken,
      pageId,
      blocks: toAppend,
    })
  }
}

interface UpdateBlockOptions {
  notionToken: string
  blockId: string
  block: Block
}

export function updateBlock ({ notionToken, blockId, block }: UpdateBlockOptions) {
  const convertedBlock = convertBlockToOutgoingBlock(block)

  return makeRequest<void>({
    notionToken,
    method: 'patch',
    path: `blocks/${blockId}`,
    body: convertedBlock,
  })
}

interface UpdatePageOptions {
  notionToken: string
  pageId: string
  properties: {
    [key: string]: {
      date: {
        start: string
      }
    }
  }
}

export function updatePage ({ notionToken, pageId, properties }: UpdatePageOptions) {
  return makeRequest<void>({
    notionToken,
    body: { properties },
    method: 'patch',
    path: `pages/${pageId}`,
  })
}

interface QueryDatabasesOptions {
  notionToken: string
  databaseId: string
}

export function queryDatabases ({ notionToken, databaseId }: QueryDatabasesOptions) {
  return makeRequest<ListDatabasesResponse>({
    notionToken,
    method: 'post',
    path: `databases/${databaseId}/query`,
  })
}

export function convertNotionBlockToOwnBlock (block: BlockObjectResponse): NotionBlock {
  const type = block.type

  return {
    // @ts-ignore
    content: clone(block[type] as Content),
    has_children: block.has_children,
    id: block.id,
    type,
  }
  // don't understand why this fails converage
  /* c8 ignore next */
}

type OutgoingBlock = UpdateBlockBodyParameters & {
  object: string
  type: BlockObjectResponse['type']
}

export function convertBlockToOutgoingBlock (block: Block): OutgoingBlock {
  return {
    object: 'block',
    type: block.type,
    [block.type]: block.content,
  }
  // don't understand why this fails converage
  /* c8 ignore next */
}

interface MakeBlockOptions {
  annotations?: Partial<RichTextItemResponse['annotations']>
  children?: OwnBlock[] | NotionBlock[]
  text: string
  type: BlockObjectResponse['type']
}

export function makeBlock ({ text, type = 'paragraph', annotations, children }: MakeBlockOptions) {
  return {
    type,
    children,
    content: {
      rich_text: [
        {
          type: 'text',
          text: {
            content: text,
          },
          annotations,
          plain_text: text,
        },
      ],
    },
  } as OwnBlock
}

export function richTextToPlainText (richText: RichTextItemResponse[]) {
  return richText
  .map(({ plain_text }) => plain_text)
  .join('')
  .trim()
}

export function getBlockPlainText (block: Block) {
  const richText = getRichText(block)

  if (!richText) return

  return richTextToPlainText(richText)
}

interface RichTextItem {
  text: {
    content: string
    link?: { url: string } | null
  }
  type?: 'text'
  annotations?: {
    bold?: boolean
    italic?: boolean
    strikethrough?: boolean
    underline?: boolean
    code?: boolean
    color?: ApiColor
  }
}

export function makeTextPart (content: string, color?: ApiColor) {
  const textPart = { text: { content } } as RichTextItem

  if (color) {
    textPart.annotations = { color }
  }

  return textPart
}

export function getMonthNameFromIndex (monthIndex: number, short = false) {
  return dayjs().month(monthIndex).format(short ? 'MMM' : 'MMMM')
}

export function getMonths ({ short }: { short?: boolean } = {}) {
  return Array.from(new Array(12)).map((_, i) => {
    return getMonthNameFromIndex(i, short)
  })
}

/* c8 ignore start */
const getDisplayPrefix = (block: NotionBlock) => {
  switch (block.type) {
    case 'paragraph': return ''
    case 'heading_1': return '# '
    case 'heading_2': return '## '
    case 'heading_3': return '### '
    case 'to_do': return `[${'checked' in block.content && block.content.checked ? '✓' : ' '}] `
    case 'bulleted_list_item': return '• '
    case 'numbered_list_item': return '1. '
    case 'toggle': return '> '
    case 'code': return '<> '
    case 'quote': return '| '
    case 'callout': return '" '
    default: return ''
  }
}

function getDisplayText (block: NotionBlock) {
  if (block.type === 'child_page') {
    return `[${'title' in block.content ? block.content.title : '<untitled>'}]`
  }

  if (block.type === 'divider') {
    return '-----'
  }

  if (getRichText(block)) {
    return `${getDisplayPrefix(block)}${getBlockPlainText(block)}`
  }

  return `(${block.type})`
}

// displays block for logging/debugging purposes
export function displayBlocks (blocks: NotionBlock[]) {
  blocks.forEach((block) => {
    debug(`[${block.id}] ${getDisplayText(block)}${block.has_children ? ' (parent)' : ''}`)
  })
}
/* c8 ignore stop */
