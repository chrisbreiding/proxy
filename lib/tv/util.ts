import type { Method } from 'axios'
import { debug, debugVerbose } from '../util/debug'
import { getEnv } from '../util/env'
import { request } from '../util/network'

const apikey = getEnv('THETVDB_API_KEY')
const pin = getEnv('THETVDB_PIN')
export const baseUrl = 'https://api4.thetvdb.com'

interface LoginResult {
  status: string
  data: {
    token: string
  }
}

async function authenticate () {
  debugVerbose('authenticate')

  try {
    const { data, status } = await request({
      method: 'post',
      url: `${baseUrl}/v4/login`,
      body: { apikey, pin },
    }) as LoginResult

    if (!data?.token) {
      throw new Error(`Could not authenticate with TheTVDB, status: ${status}`)
    }

    return data.token as string
  } catch (error: any) {
    debug('Authentication failed:', error?.stack || error)

    throw error
  }
}

interface MakeRequestOptions {
  method?: Method
  params?: { [key: string]: string }
  path: string
  token?: string
}

export async function makeRequest (options: MakeRequestOptions) {
  const { method, params, path } = options
  const token = options.token || await authenticate()
  const url = `${baseUrl}/v4/${path}`

  debugVerbose('request: %o', { method, url, params, token })

  // TODO: handle error
  const result = await request({
    headers: {
      'Authorization': `Bearer ${token}`,
    },
    method,
    params,
    url,
  })

  result.token = token

  return result
}
