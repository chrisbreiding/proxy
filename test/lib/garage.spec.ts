import { readJsonSync } from 'fs-extra'
import mockFs from 'mock-fs'
import nock from 'nock'
import path from 'path'
import { afterAll, beforeEach, describe, expect, it, TestContext } from 'vitest'

import { handleServer, replaceStackLines } from '../util'
import { startServer } from '../../index'
import { getGarageData, GarageState, PersistentDataStructure } from '../../lib/garage'

process.env.API_KEY = 'key'

const defaultData = {
  left: 'closed' as const,
  right: 'closed' as const,
}

function mockData (data: PersistentDataStructure | null) {
  mockFs({
    'data': {
      'garage-data.json': JSON.stringify(data),
    },
    'views': mockFs.load(path.resolve(process.cwd(), 'views')),
  })
}

function mockError () {
  mockFs({
    'data': {},
    'views': mockFs.load(path.resolve(process.cwd(), 'views')),
  })
}

function getData () {
  return readJsonSync('./data/garage-data.json')
}

describe('lib/garage', () => {
  handleServer(startServer)

  beforeEach(() => {
    nock.cleanAll()
  })

  afterAll(() => {
    mockFs.restore()
  })

  describe('GET /garage-states/:key', () => {
    it('returns garage states from persistent data', async (ctx) => {
      mockData(defaultData)

      const res = await ctx.request.get('/garage-states/key')

      expect(res.status).to.equal(200)
      expect(res.body).to.deep.equal(defaultData)
    })

    it('returns unknown states if no data', async (ctx) => {
      mockData(null)

      const res = await ctx.request.get('/garage-states/key')

      expect(res.status).to.equal(200)
      expect(res.body).to.deep.equal({
        left: 'unknown',
        right: 'unknown',
      })
    })

    it('returns unknown states if error', async (ctx) => {
      mockError()

      const res = await ctx.request.get('/garage-states/key')

      expect(res.status).to.equal(200)
      expect(res.body).to.deep.equal({
        left: 'unknown',
        right: 'unknown',
      })
    })

    it('status 403 if key does not match', async (ctx) => {
      const res = await ctx.request.get('/garage-states/nope')

      expect(res.status).to.equal(403)
    })
  })

  describe('POST /garage-states/:door/:state/:key', () => {
    function nockNotify (message: string, optional = false) {
      process.env.IFTTT_WEBHOOK_KEY = 'iftttkey'

      return nock('https://maker.ifttt.com')
      .get(`/trigger/notify/with/key/iftttkey?value1=${encodeURIComponent(message)}`)
      .optionally(optional)
      .reply(200)
    }

    async function assertNoNotifications (ctx: TestContext, states: PersistentDataStructure, newState: GarageState) {
      const notifyFailed = nockNotify('The left garage door failed to close!')
      const notifyOpened = nockNotify('The left garage door opened')

      mockData(states)

      await ctx.request.post(`/garage-states/left/${newState}/key`)

      expect(notifyFailed.isDone(), 'notification was sent').to.be.false
      expect(notifyOpened.isDone(), 'notification was sent').to.be.false

      nock.abortPendingRequests()
    }

    it('returns an empty object', async (ctx) => {
      mockData(defaultData)

      const res = await ctx.request.post('/garage-states/left/open/key')

      expect(res.status).to.equal(200)
      expect(res.body).to.deep.equal({})
    })

    it('handles empty persistent data', async (ctx) => {
      mockData(null)

      const res = await ctx.request.post('/garage-states/left/open/key')

      expect(res.status).to.equal(200)
      expect(res.body).to.deep.equal({})
    })

    it('returns an empty object if error', async (ctx) => {
      mockError()

      const res = await ctx.request.post('/garage-states/left/open/key')

      expect(res.status).to.equal(200)
      expect(res.body).to.deep.equal({})
    })

    it('sets garage states in persistent data', async (ctx) => {
      mockData(defaultData)

      await ctx.request.post('/garage-states/left/open/key')

      expect(getData()).to.deep.equal({
        left: 'open',
        right: 'closed',
      })
    })

    it('notifies if new and previous states are open', async (ctx) => {
      const notifyFailed = nockNotify('The left garage door failed to close!')

      mockData({
        left: 'open',
        right: 'open',
      })

      await ctx.request.post('/garage-states/left/open/key')

      expect(notifyFailed.isDone(), 'notification was not sent').to.be.true
    })

    it('notifies if open and notifyOnOpen: true', async (ctx) => {
      const notifyOpened = nockNotify('The left garage door opened')

      mockData({
        left: 'closed',
        right: 'open',
        notifyOnOpen: true,
      })

      await ctx.request.post('/garage-states/left/open/key')

      expect(notifyOpened.isDone(), 'notification was not sent').to.be.true
    })

    it('does not notify if new and previous states are closed', async (ctx) => {
      await assertNoNotifications(ctx, {
        left: 'closed',
        right: 'closed',
      }, 'closed')
    })

    it('does not notify if new and previous states are different', async (ctx) => {
      await assertNoNotifications(ctx, {
        left: 'closed',
        right: 'closed',
      }, 'open')
    })

    it('does not notify if state is open and notifyOnOpen is false', async (ctx) => {
      await assertNoNotifications(ctx, {
        left: 'closed',
        right: 'closed',
        notifyOnOpen: false,
      }, 'open')
    })

    it('ignores notify errors', async (ctx) => {
      process.env.IFTTT_WEBHOOK_KEY = 'iftttkey'

      nock('https://maker.ifttt.com')
      .get(`/trigger/notify/with/key/iftttkey?value1=${encodeURIComponent('The left garage door opened')}`)
      .reply(500)

      mockData({
        left: 'closed',
        right: 'open',
        notifyOnOpen: true,
      })

      await ctx.request.post('/garage-states/left/open/key')
    })
  })

  describe('POST /garage/notify-on-open/:notifyOnOpen/:key', () => {
    it('sets notifyOnOpen in persistent data', async (ctx) => {
      mockData(defaultData)

      await ctx.request.post('/garage/notify-on-open/true/key')

      expect(getData()).to.deep.equal({
        left: 'closed',
        right: 'closed',
        notifyOnOpen: true,
      })

      await ctx.request.post('/garage/notify-on-open/false/key')

      expect(getData()).to.deep.equal({
        left: 'closed',
        right: 'closed',
        notifyOnOpen: false,
      })
    })

    it('handles empty persistent data', async (ctx) => {
      mockData(null)

      const res = await ctx.request.post('/garage/notify-on-open/true/key')

      expect(res.status).to.equal(200)
      expect(res.body).to.deep.equal({})
    })

    it('returns empty object if error', async (ctx) => {
      mockError()

      const res = await ctx.request.post('/garage/notify-on-open/true/key')

      expect(res.status).to.equal(200)
      expect(res.body).to.deep.equal({})
    })

    it('status 403 if key does not match', async (ctx) => {
      const res = await ctx.request.post('/garage/notify-on-open/true/nope')

      expect(res.status).to.equal(403)
    })
  })

  describe('GET /garage/:key', () => {
    it('renders garage view', async (ctx) => {
      mockData({
        left: 'closed',
        right: 'open',
      })

      const res = await ctx.request.get('/garage/key')

      expect(res.status).to.equal(200)
      expect(res.headers['content-type']).to.equal('text/html; charset=utf-8')

      mockFs.restore()
      expect(res.text).toMatchSnapshot()
    })

    it('handles empty persistent data', async (ctx) => {
      mockData(null)

      const res = await ctx.request.get('/garage/key')

      expect(res.status).to.equal(200)
      expect(res.headers['content-type']).to.equal('text/html; charset=utf-8')

      mockFs.restore()
      expect(res.text).toMatchSnapshot()
    })

    it('renders error view if error', async (ctx) => {
      mockError()

      const res = await ctx.request.get('/garage/key')

      expect(res.status).to.equal(200)
      expect(res.headers['content-type']).to.equal('text/html; charset=utf-8')

      mockFs.restore()
      expect(replaceStackLines(res.text)).toMatchSnapshot()
    })

    it('status 403 if key does not match', async (ctx) => {
      const res = await ctx.request.get('/garage/nope')

      expect(res.status).to.equal(403)
    })
  })

  describe('#getGarageData', () => {
    it('returns persistent data', async () => {
      mockData(defaultData)

      const res = await getGarageData()

      expect(res).to.deep.equal(defaultData)
    })
  })
})
