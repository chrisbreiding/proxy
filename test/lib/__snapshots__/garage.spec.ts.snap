// Vitest Snapshot v1

exports[`lib/garage > GET /garage/:key > renders error view if error 1`] = `
"<!DOCTYPE html>
<html>
<head>
  <meta charset=\\"utf-8\\">
  <title>Server Error</title>
  <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge,chrome=1\\">
  <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
  <style>
    html {
      background: #333;
    }
    body {
      color: #f8f8f8;
      font-family: sans-serif;
      padding: 2em;
    }

    p, pre {
      font-size: 20px;
    }
  </style>
</head>
<body>
  <h1>An Error Occurred</h1>
  <h2>Error Message</h2>
  <p>ENOENT, no such file or directory &#x27;./data/garage-data.json&#x27;</p>
  <h2>Error Stack</h2>
  <pre>Error: ENOENT, no such file or directory &#x27;./data/garage-data.json&#x27;
    [stack lines]</pre>
</body>
</html>"
`;

exports[`lib/garage > GET /garage/:key > renders garage view 1`] = `
"<!DOCTYPE html>
<html>
<head>
  <meta charset=\\"utf-8\\">
  <title>Garage States</title>
  <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge,chrome=1\\">
  <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
  <style>
    html {
      background: #333;
    }
    body {
      color: #f8f8f8;
      display: flex;
      font-family: sans-serif;
      justify-content: center;
      padding-top: 2em;
    }
    .state {
      align-items: center;
      display: flex;
      flex-direction: column;
      justify-content: center;
      margin: 1em 0.4em;
      min-width: 140px;
    }
    .state-icon {
      background-color: blue;
      height: 100px;
      -webkit-mask-size: 100px;
      mask-size: 100px;
      width: 100px;
    }
    .state-icon-open { background-color: #FDA4A3 }
    .state-icon-closed { background-color: #E5FFB2 }
    .state-icon-unknown { background-color: #E6E6E6 }
    .state p {
      text-align: center;
      margin: 0.5em 0 0;
    }
  </style>
</head>
<body>
  <div class=\\"state\\">
    <div
      class=\\"state-icon state-icon-closed\\"
      style=\\"mask-image: url(/garage-closed.png); -webkit-mask-image: url(/garage-closed.png);\\"
    ></div>
    <p>Left: closed</p>
  </div>
  <div class=\\"state\\">
    <div
      class=\\"state-icon state-icon-open\\"
      style=\\"mask-image: url(/garage-open.png); -webkit-mask-image: url(/garage-open.png);\\"
    ></div>
    <p>Right: open</p>
  </div>
</body>
</html>"
`;
